### ISSUE TITLE
A "safe" subsection of the library

### ISSUE THREAD

**Issue by MatthewDaggitt** [agda/agda-stdlib#143]
As discussed in #122 it is currently impossible to typecheck a development relying on the standard library with the --safe option because it does not isolate properly its uses of unsafe features.

The agreed upon course of action is that there will be a Safe.* subset of modules. The current module names will be kept the same and will re-export exactly what they currently contain, ensuring full backwards compatibility whilst making it possible to only rely on the subset that is recognised as being safe by Agda.

**Comment by gallais** [comment]
I have come across something fairly nasty: [`Agda.Builtin.Int`](https://agda.github.io/agda-stdlib/Agda.Builtin.Int.html) depends upon [`Agda.Builtin.String`](https://agda.github.io/agda-stdlib/Agda.Builtin.String.html) which uses a `postulate` to declare the `String` type. As a consequence any module importing `Agda.Builtin.Int` cannot be accepted by the current `--safe` option Agda has.

In other words either all `Data.Integer.*` and any other module depending upon them have to be in the unsafe part of the library or Agda needs to provide e.g. `Safe.Agda.Builtin.Int` without `primShowInteger` (and therefore without any dependence on `Agda.Builtin.String`).

**Comment by UlfNorell** [comment]
It sounds like maybe we should relax the requirements of `--safe` to accept postulates of type `Set`.

**Comment by gallais** [comment]
I don't understand why [Agda.Primitive](https://agda.github.io/agda-stdlib/Agda.Primitive.html) doesn't trigger an error given that it does use `postulate` all over the place.

**Comment by gallais** [comment]
All the tests are green. I think the branch is more or less ready to be merged (and I'd like to avoid it diverging too much from `master` before merge because that's a lot of changes to maintain!). When / If Agda gets more liberal in terms of what it accepts as safe, we can move more things into `Safe/`. It should all be transparent from the users' point of view anyway.

The main issues I see are:

## Wider Use

Do we want to ask people on the agda ML to try to check their developments with the current `safe` branch to make sure it doesn't break anything (e.g. we haven't forgotten to re-export any names anywhere)?

Or would it be enough to run Agda's stdlib-based test suite on the modified one? In which case I guess it's simply a matter of creating a dummy branch on agda/agda where we update the submodule to point to the safe branch.

## Stylistic Issue

Do we want `{-# OPTIONS --safe #-}` as a header in all of the `Safe/` modules or are we just happy with having travis check `Safe/index.agda` with the appropriate command line options? At the moment we have both the header and the travis separate check (the latter is necessary even though we have the headers because of agda/agda#2487).

**Comment by MatthewDaggitt** [comment]
Just had a look at the branch. What's the plan for maintaining this? Is every safe change to the non-safe part going to have to be copied over to the safe part manually?

**Comment by gallais** [comment]
The idea is to replace `master` with `safe`. Given that each unsafe module re-exports the content of its safe counterpart, it should be completely transparent for end users.

New definitions can be added either to the safe module if they are innocuous, or to the unsafe one if they're not.

**Comment by MatthewDaggitt** [comment]
Oh yes. Apologies, hadn't noticed that all the non-safe versions contents had been moved.

> Given that each unsafe module re-exports the content of its safe counterpart, it should be completely transparent for end users.

Until they go browsing through it! 

I'm a little worried about what this is going to do to library usability. For instance whereas before looking for the decidability of the divides relation `_∣_`, would involve only looking through `Data.Nat.Divisibility`. However it is now going to require browsing both `Safe.Data.Nat.Divisibility` and `Data.Nat.Divisibility`. What's more you can't off-hand predict which one it's going to belong to.

This seems like a big impact on the average user who's not that concerned about rigorous safety. Can users who are concerned about safety not just type-check it on the command line using the `--safe` option? Won't that automatically highlight the unsafe terms used? Or does it over-estimate, and mark otherwise safe terms unsafe if they share a file with unsafe terms?

**Comment by gallais** [comment]
> However it is now going to require browsing both Safe.Data.Nat.Divisibility and Data.Nat.Divisibility which are deep in different parts of the heirarchy. What's more you can't off-hand predict which one it's going to belong to.

The doc for the unsafe module systematically links to the safe one because it imports it.

> Can users who are concerned about safety not just type-check it on the command line using the --safe option?

`--safe` only succeeds if all of the definitions in a module (and all the module it depends upon) are safe. Hence the need for a split: it is currently near impossible to typecheck a development relying on the standard library with the `--safe` option. I'm actually surprised that after 100+ articles using Agda published, no reviewer seems to have complained about this.

**Comment by MatthewDaggitt** [comment]
> The doc for the unsafe module systematically links to the safe one because it imports it.

Hmm... I guess so.

> --safe only succeeds if all of the definitions in a module (and all the module it depends upon) are safe. 

Whereas what it should ideally do is only check the definitions in other modules that are actually __used__.  From my position of complete ignorance, this seems like it should be doable using some sort of flow analysis. I'm not sure if it is or not! If it is possible, wouldn't this be a far better solution than partitioning the standard library in two?

I guess its kind of the dual to agda/agda#2487. File options should be more strict with respect to imported definitions, and command line options need to be less strict....

**Comment by gallais** [comment]
> Whereas what it should ideally do is only check the definitions in other modules that are actually used. 

But then you'd have to call `agda --safe` on every single module attached to a paper to make sure the whole development is not using any dirty tricks. Right now you can simply generate an `Everything.agda` and typecheck that.

**Comment by MatthewDaggitt** [comment]
> But then you'd have to call agda --safe on every single module attached to a paper to make sure the whole development is not using any dirty tricks

Aren't you going to have to do that anyway? Even if you've only used code from the `Safe` hierarchy in the standard library, you still need to check that your own code doesn't use any dirty tricks.

**Comment by gallais** [comment]
You can call `agda --safe` on a module `Everything.agda` importing all the modules in your development and it'll make sure everything is a-ok. Just like the [travis test](https://travis-ci.org/agda/agda-stdlib/builds/260086278#L2819) at the moment checks the whole of `Safe/` simply by checking `Safe/index.agda`.

**Comment by MatthewDaggitt** [comment]
Oh okay, that makes sense. Thank you for your patience and taking the time to explain it to me! I have experience with exactly one agda development and that's my own so I'm not fully versed in all the tricks...

Okay well then perhaps there should be a `--safe-narrow` and `--safe-broad` option to cover both options. It just seems weird that a module that only uses safe definitions should be labelled as unsafe just because some of its dependencies happen to share a file with unsafe definitions. In fact the same thing should probably apply to other options like `--without-k`.

**Comment by gallais** [comment]
> Okay then perhaps there should be a --safe-narrow and --safe-broad option to cover both options

Won't the problem still be the same? You'll want `--safe-narrow` on all of the modules corresponding to a paper (ALL of the theorems have to be valid) but if it relies on the standard library, you'll need that constraint to magically change to `--safe-broad` whenever definitions rely upon that dependency.

**Comment by MatthewDaggitt** [comment]
I have to confess I was imagining a scenario where issue agda/agda#2487 was fixed as well. Then it'll suffice to simply have the `--safe-broad` pragma at the top of each of your top level theorem files.

**Comment by fredrikNordvallForsberg** [comment]
> I don't understand why [Agda.Primitive](https://agda.github.io/agda-stdlib/Agda.Primitive.html) doesn't trigger an error given that it does use `postulate` all over the place.

Perhaps more of a side remark, but this is because safe mode is [explicitly turned off]( https://github.com/agda/agda/blob/master/src/full/Agda/Interaction/Imports.hs#L242-L255) when `Agda.Primitive` is typechecked.

**Comment by jespercockx** [comment]
> Whereas what it should ideally do is only check the definitions in other modules that are actually used. From my position of complete ignorance, this seems like it should be doable using some sort of flow analysis. I'm not sure if it is or not! If it is possible, wouldn't this be a far better solution than partitioning the standard library in two?

Maybe it would make sense if Agda internally marks definitions that *could* have been checked with `--safe` as such, even if the flag is not enabled. That way, when importing a file without --safe, Agda can check for the definitions you actually use whether they are safe. Although this would probably require quite some refactoring of the other flags that are excluded by `--safe`.

**Comment by nad** [comment]
> The agreed upon course of action is that there will be a Safe.* subset of modules.

I think I'd prefer `M.Safe` to `Safe.M`: In my opinion the word `Safe` is more noisy at the start than at the end. Furthermore `M` and `M.Safe` are close to each other, both alphabetically and in terms of the directory hierarchy.

**Comment by HuStmpHrrr** [comment]
Hi people, I came around here from another similar issue #255. Pretty much the same request, though I don't get why the basic assumption is the language is unsafe? I'd prefer the other way since all definitions should come safe in the first place (except those unsafe by definition), and only after then we have unsafe stuff around the beautiful utopia?

**Comment by gallais** [comment]
Is this worth being an 1.0 kind of issue? This would induce a big backwards-compatibility breakage but we would also have the guarantee it is going to be canon in the further releases.

**Comment by MatthewDaggitt** [comment]
I still feel that many of the current issues could be circumvented by making `safe` more discriminating and detect that there are no unsafe terms used, even if they live in modules that do contain unsafe terms (see agda/agda#2789).

I realise there's strong demand to get around its limitations in the mean time. I agree with @HuStmpHrrr , I far prefer the style of `X.Y.Z.Unsafe` over duplicating the entire hierarchy and having `Safe.X.Y.Z`. Things should be safe by default, and we only need to make people aware when they're using unsafe features. @gallais I know you put a lot of work into the `safe` branch. What do you think? 


> Is this worth being an 1.0 kind of issue? This would induce a big backwards-compatibility breakage but we would also have the guarantee it is going to be canon in the further releases.

I think so too. I have nearly completed all of my planned structural re-organisations. The last three are:
- Membership (#94 & #155) 
- PartialOrderReasoning (see #185 )
- Get rid of `Fin.Dec`
After that I don't anticipate any more major structural changes, so it seems like it would be a good time to at long last release version 1.0!

**Comment by HuStmpHrrr** [comment]
I was thinking about it for the last couple days, and I hope I got a better picture of this.

I think it might not be the problem of this library (though it's indeed very nice if unsafe code can be extracted to submodules in the previously mentioned way). I do agree that it's going to be much better of, if the compiler can check the safety of every function, like @MatthewDaggitt suggested. One of the very important aspects of being able to do so, is to make tactics more realizable. Since tactics is based on yet another bunch of postulates, there is no way if a program is written by tactics and can be safe. However, the program built by tactics itself can be entirely safe; while on the other hand, one can indeed embed unsafe code via tactics as well.

From that aspect, I believe adding such functionalities to the compiler will help largely to eliminate boring mechanical programming in agda, and that also means that it's more on what the compiler does, instead of what this particular library looks like.

**Comment by asr** [comment]
> Why doesn't the command:
>  ```
>  agda --safe src/Relation/Binary/HeterogeneousEquality/Quotients.agda
>  ``` 
 > throw an error? 


When trying of reproducing the problem in Agda `stable-2.5` branch I got the error:

```
$ make up-to-date-std-lib
...
cabal exec -- GenerateEverything
GenerateEverything: src/Relation/Binary/HeterogeneousEquality/Quotients.agda is malformed.
CallStack (from HasCallStack):
  error, called at GenerateEverything.hs:67:15 in main:Main
GNUmakefile:16: recipe for target 'Everything.agda' failed
```

**Comment by MatthewDaggitt** [comment]
Hmm weird. I'm on a clean version of 2.5.3. I'm having the same problem with the `--without-K` option...

**Comment by MatthewDaggitt** [comment]
Tracked it down, it was another subtler instance of agda/agda#2487. As I'd compiled the whole library without `--safe`, the `.agdai` files were lying to me when I subsequently recompiled it all with `--safe`. I'm rapidly coming to the opinion that the `--safe` option currently doesn't provide anything much in the way of guarantees (either in-file or at the command line) :disappointed:

**Comment by gallais** [comment]
`--safe` + `--ignore-interfaces` is pretty reliable

**Comment by WolframKahl** [comment]
On Thu, Apr 12, 2018 at 07:46:35AM -0700, G. Allais wrote:
> `--safe` + `--ignore-interfaces` is pretty reliable

Except that it cannot be resumed easily if there is a crash
before completion
(most likely due to starting Agda with insufficient heap space,
 which will in turn quite likely be due to hardware constraints...
).
Admittadly, this should not be a problem for the standard library.

But in general, removing all .agdai files is safer.

**Comment by nad** [comment]
> I'm rapidly coming to the opinion that the `--safe` option currently doesn't provide anything much in the way of guarantees (either in-file or at the command line)

The `--safe` feature was originally implemented by someone who is no longer contributing to Agda, and it seems as if no regular developer is eager to maintain it. Fixing the documented bugs related to `--safe` requires some work, but I think it should be doable for someone who is not intimately familiar with the internals of Agda. Contributions are welcome.

**Comment by MatthewDaggitt** [comment]
>>I'm rapidly coming to the opinion that the --safe option currently doesn't provide anything much in the way of guarantees (either in-file or at the command line)
>
> Fixing the documented bugs related to --safe requires some work, but I think it should be doable for someone who is not intimately familiar with the internals of Agda. Contributions are welcome.

While I would love to, I feel I already spend far too much of my time of my PhD on internet routing protocols playing around with Agda! Sorry, I realise that my comment perhaps came off as a complaint, it wasn't meant to be.

**Comment by MatthewDaggitt** [comment]
I've started working on creating the various `Unsafe` modules (very similar to @gallais work on #122). It's all mostly okay (although haven't got to `Data.Integer` yet...) but the first problem I've come across is in `Data.Nat.DivMod`.

https://github.com/agda/agda-stdlib/blob/5914224fe2ea9a329286fb6f0f6216a93dc59f36/src/Data/Nat/DivMod.agda#L29-L35

The problem is that the `remainder` field is of type `Fin divisor` and not `ℕ`. For efficiency's sake (although apparently it doesn't work very well, see #150) we erase all the various proof that allows us to convert from `ℕ` to `Fin divisor`. The erasure uses `trustMe` which is of course, unsafe. This then pollutes large parts of the following modules that depend on `DivMod`:
```
Data.Nat.Divisibility
Data.Nat.Primality
Data.Nat.Show
Data.Digit
```
There are two options that I see:

1. Proceed as planned and stick all dependencies in `Unsafe` modules. 
    * **Advantages**: A little more backwards compatible (although obviously all module imports will have to change). 
    * **Disadvantages**: we label large sections of our `Nat` library as `Unsafe`, including our primality test which is tad embarrassing. 
2. Change the type of `_mod_` and the field `remainder` to `ℕ` rather than `Fin divisor`, and then create separate proofs that show they are all related in the right way.
    * **Advantages**: This would provide a faster and safe version of `_mod_`, and hence avoid polluting the safety of the rest of the library with the use of `trustme`. The proofs that the right properties hold, still exist, and users themselves can make the choice of whether or not to erase them for efficiency reasons.
    * **Disadvantages**: not backwards compatible

I'd prefer the 2nd option as it feels much more principled. Not all users will require the proofs, but those that do then have much finer grained control over how they use them. The non-backwards compatability is unfortunate but that's why we're labelling it v1.0?

**Comment by HuStmpHrrr** [comment]
> Disadvantages: Performance gets even worse.

I am astonished by the performance part. Might I ask how important performance is, and how many people are relying on it? Sorry about my ignorance...

What about irrelevance? I've not seen the description in official doc, but from my understanding, this should be the secrete magic to tame the tradeoff between complex proof flow and performance, is it not?

**Comment by MatthewDaggitt** [comment]
> I am astonished by the performance part. Might I ask how important performance is, and how many people are relying on it? Sorry about my ignorance...

No idea. I don't personally normalise any of my proofs. But given that Ulf has put significant effort into building [agda-prelude](https://github.com/UlfNorell/agda-prelude) with a strong focus on performance I gather it's definitely important for some people.  Also issue #150 is is actually directly related to the performance of this part of the library.

> What about irrelevance? I've not seen the description in official doc, but from my understanding, this should be the secrete magic to tame the tradeoff between complex proof flow and performance, is it not?

Again, as I never worry about actually normalising I've never had any need to dip into irrelevance. Maybe  someone with more experience could chime in?

Overall, I think the consensus is that the standard library focuses on making proofs easier rather than on performance. Having said that, we should probably generally avoid making performance worse except for good reason. Hopefully we can come to some conclusion here whether or not safety is a good enough reason!

**Comment by MatthewDaggitt** [comment]
#350 which has just been merged provides a way forward on this problem. It adds a new "safe" modular arithmetic library. The old library can then be placed in the `Unsafe` module. The dependencies have been updated to use the new library.

**Comment by MatthewDaggitt** [comment]
Closed with 8885d7a
